File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        Queue<int[]> queue = new LinkedList<>();
        int[][] directions = new int[][]{{0,1},{1,0},{0,-1},{-1,0}};
        int res = 0;
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(grid[i][j]==1){
                    grid[i][j]=0;
                    queue.offer(new int[] {i,j});
                    res = Math.max(res,count_max_area(grid, queue, directions));
                }
            }
        }
        return res;
    }

    public int count_max_area(int[][] grid, Queue<int[]> queue, int[][] directions){
        int count_area = 1;
        while(!queue.isEmpty()){
            int[] rowcol = queue.poll();
            for(int[] dir:directions){
                int new_r = rowcol[0]+dir[0];
                int new_c = rowcol[1]+dir[1];

                if(new_r>=0 && new_r<grid.length && new_c>=0 && new_c<grid[0].length){
                    if(grid[new_r][new_c]==1){
                        count_area++;
                        grid[new_r][new_c]=0;
                        queue.add(new int[] {new_r,new_c});
                    }
                }
            }
        }
        return count_area;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[str]]) -> int:
        max_area = 0
        directions = [(1,0),(0,1),(-1,0),(0,-1)]
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]==1:
                    grid[i][j] = 0
                    get_area = self.count_island(directions, i, j, grid)
                    max_area = max(max_area, get_area)
        return max_area
    def count_island(self, directions, i, j, grid):
        stack = []
        stack.append((i,j))
        count_island = 1
        while stack:
            r, c = stack.pop()
            for row, col in directions:
                new_row = row+r
                new_col = col+c
                if 0<=new_row<len(grid) and 0<=new_col<len(grid[0]):
                    if grid[new_row][new_col]==1:
                        count_island+=1
                        grid[new_row][new_col] = 0
                        stack.append((new_row,new_col))
        return count_island